File: src/matrix/matrix.zig
const std = @import("std");
const print = std.debug.print;
const rand = @import("std").rand;
const time = @import("std").time;

const M: i64 = 800; // Adjust the matrix size to increase memory usage
const N: i64 = M;


// Function to generate a random matrix
fn generateRandomMatrix(matrix: *[M][N]i64, prng: *rand.DefaultPrng) void {
  for (matrix, 0..) |row, i| {
    for (row, 0..) |_, j| {
      matrix[i][j] = prng.random().intRangeAtMost(i64, 1, 100000);
    }
  }
}

// Function for matrix multiplication
fn matrixMultiplication(result: *[M][N]i64, a: *[M][N]i64, b: *[M][N]i64) void {
  for (result, 0..) |row, i| {
    for (row, 0..) |_, j| {
      result[i][j] = 0;
    }
  }
  
  for (result, 0..) |row, i| {
    for (row, 0..) |_, j| {
      for (0..N) |k| {
        result[i][j] = result[i][j] + a[i][k] * b[k][j];
      }
    }
  }
}

pub fn main() !void {
  const startTime = time.milliTimestamp();

  print("matrixSize: {}\n", .{M});

  var prng = std.rand.DefaultPrng.init(blk: {
    var seed: u64 = undefined;
    try std.os.getrandom(std.mem.asBytes(&seed));
    break :blk seed;
  });
  
  var a: [M][N]i64 = undefined;
  var b: [M][N]i64 = undefined;
  var result: [M][N]i64 = undefined;
  
  generateRandomMatrix(&a, &prng);
  generateRandomMatrix(&b, &prng);
  matrixMultiplication(&result, &a, &b);
  
  const endTime = time.milliTimestamp();
  const elapsedTime = endTime - startTime;
  
  print("Matrix Multiplication time: {}ms\n", .{elapsedTime});
}

File: src/matrix/matrix.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Function to generate a random matrix
void generateRandomMatrix(int rows, int cols, double **matrix) {
    for (int i = 0; i < rows; i++) {
        matrix[i] = (double *)malloc(cols * sizeof(double));
        for (int j = 0; j < cols; j++) {
            matrix[i][j] = ((double)rand() / RAND_MAX);
        }
    }
}

// Function for matrix multiplication
void matrixMultiplication(int rowsA, int colsA, double **a, int rowsB, int colsB, double **b, double **result) {
    if (colsA != rowsB) {
        fprintf(stderr, "Incompatible matrices for multiplication\n");
        exit(1);
    }

    for (int i = 0; i < rowsA; i++) {
        result[i] = (double *)malloc(colsB * sizeof(double));
        for (int j = 0; j < colsB; j++) {
            result[i][j] = 0;
            for (int k = 0; k < colsA; k++) {
                result[i][j] += a[i][k] * b[k][j];
            }
        }
    }
}

int main() {
    clock_t start = clock();

    int matrixSize = 1000; // Adjust the size of the matrix to increase memory usage
    printf("matrixSize: %d\n", matrixSize);

    // Generate random matrices and multiply
    double **matrixA = (double **)malloc(matrixSize * sizeof(double *));
    double **matrixB = (double **)malloc(matrixSize * sizeof(double *));
    double **resultMatrix = (double **)malloc(matrixSize * sizeof(double *));
    generateRandomMatrix(matrixSize, matrixSize, matrixA);
    generateRandomMatrix(matrixSize, matrixSize, matrixB);
    matrixMultiplication(matrixSize, matrixSize, matrixA, matrixSize, matrixSize, matrixB, resultMatrix);

    // Free allocated memory
    for (int i = 0; i < matrixSize; i++) {
        free(matrixA[i]);
        free(matrixB[i]);
        free(resultMatrix[i]);
    }
    free(matrixA);
    free(matrixB);
    free(resultMatrix);

    clock_t end = clock();
    double cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;
    printf("Matrix Multiplication time: %fs\n", cpu_time_used);

    return 0;
}